tests/std/tests/GH_001017_discrete_distribution_out_of_range/bad_random_engine.hpp

```diff
@@ -0,0 +1,192 @@
+            current_value_ = (current_value_ & lower_bits) << 1 | (current_value_ & top_bit) >> (Width - 1);
+
+            if (current_shift_ < Width - 1 && current_bit_count_ != 0 && current_bit_count_ != Width) {
+                ++current_shift_;
+                return *this;
+            }
+
+            current_shift_ = 0;
+
+            if (current_bit_count_ < final_bit_count) { // n 1's -> n 0's
+                current_bit_count_ = Width - current_bit_count_;
+                current_value_ ^= mask_bits;
+            } else if (current_bit_count_ > final_bit_count) { // n 0's -> (n+1) 1's
+                current_bit_count_ = Width - current_bit_count_ + 1;
+                current_value_     = (current_value_ ^ mask_bits) << 1 | value_type{1};
+            } else { // all bit patterns have been generated, back to all 0's
+                current_bit_count_ = 0;
+                current_value_     = value_type{0};
+            }
+
+            return *this;
+        }
+
+        constexpr my_iter operator++(int) noexcept {
+            const my_iter old = *this;
+            ++*this;
+            return old;
+        }
+
+        friend constexpr bool operator==(const my_iter& a, const my_iter& b) noexcept {
+            return *a == *b;
+        }
+
+        friend constexpr bool operator!=(const my_iter& a, const my_iter& b) noexcept {
+            return !(a == b);
+        }
+
+        friend constexpr bool operator==(const my_iter& iter, const my_sentinel sentinel) noexcept {
+            (void) sentinel;
+            return *iter == 0;
+        }
+
+        friend constexpr bool operator!=(const my_iter& iter, const my_sentinel sentinel) noexcept {
+            return !(iter == sentinel);
+        }
+
+        friend constexpr bool operator==(const my_sentinel sentinel, const my_iter& iter) noexcept {
+            return iter == sentinel;
+        }
+
+        friend constexpr bool operator!=(const my_sentinel sentinel, const my_iter& iter) noexcept {
+            return !(iter == sentinel);
+        }
+
+    private:
+        value_type current_value_ = 0;
+        int current_bit_count_    = 0;
+        int current_shift_        = 0;
+    };
+} // namespace detail
+
+template <typename UInt, int Width = std::numeric_limits<UInt>::digits, int Dimension = 1>
+class bad_random_engine {
+    // Generates bit patterns with at most two transitions between 0's and 1's.
+    // (e.g. 00000000, 11111111, 00001111, 11110000, 00011000, 11100111)
+    // When its output is grouped into subsequences of length Dimension, it cycles through all possible subsequences
+    // containing only such bit patterns. Bit patterns with few 1's or few 0's are generated first, starting from all
+    // 0's and all 1's.
+
+    static_assert(std::is_integral<UInt>::value, "bad_random_engine: UInt should be unsigned integeral type");
+    static_assert(std::is_unsigned<UInt>::value, "bad_random_engine: UInt should be unsigned integeral type");

```

Comment:
Fix comment typos. Also, MSVC supports the type traits variable templates unconditionally, so we use them everywhere (they were added in C++17, but we implemented them before Standard modes, so they're available in `/std:c++14` which actually means "C++14 plus all of the C++17 stuff we had implemented at the time, to avoid breaking code unnecessarily").
```suggestion
    static_assert(std::is_integral_v<UInt>, "bad_random_engine: UInt should be unsigned integral type");
    static_assert(std::is_unsigned_v<UInt>, "bad_random_engine: UInt should be unsigned integral type");
```